File: 122. Best Time to Buy and Sell Stock II.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit = 0;
        int bought = -1;
        int n = prices.size();
        for (int i = 0; i < n; ++i) {
            if (bought == -1 && i < n - 1 && prices[i] < prices[i + 1]) {
                bought = i;
            } else if (bought != -1 && (i == n - 1 || prices[i] > prices[i + 1])) {
                // remember to handle i == n - 1 case
                profit += prices[i] - prices[bought];
                bought = -1;
            }
        }
        
        return profit;
    }
};

File: 122. Best Time to Buy and Sell Stock II.py
# Time: O(n), space: O(1)
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        buy = -1
        profit = 0
        for i in range(len(prices) - 1):
            if prices[i] < prices[i + 1] and buy == -1:
                buy = i
            elif prices[i] > prices[i + 1] and buy != -1:
                profit += prices[i] - prices[buy]
                buy = -1
        if buy != -1: profit += prices[-1] - prices[buy] # don't forget
        return profit

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if (not prices): return 0
        buy = -prices[0]
        sell = 0
        for i, price in enumerate(prices):
            buy = max(buy, sell - price)
            sell = max(sell, buy + price)
        return sell
